### DISCUSSION TITLE
useFieldArray performance for many fields is really bad

### DISCUSSION THREAD

**Discussion by Geddard** [react-hook-form/react-hook-form#3585] [Q&A]
Hey there, I was very happy when I found the useFieldArray hook in the docs, since it solves pretty much every scenario I need to cover in my project.

But the problem I have is that I'm creating a form which has about 12 input fields (and there will be forms with even double that), and when I try to append more rows of the same form I can see the browser struggling because every previous already rendered row triggers a re-render because the form is making some internal change (that are not even reflected in the props).

So I though, I can just use React.memo() on my input field for example, which improves performance by a lot, but then when I try to remove a row, the form explodes. Here's an example of the rendering i have:

```
const formFields = useMemo(
    () => tableColumns.reduce((items, { queryParam }) => ({ ...items, ...{ [queryParam]: '' } }), {}),
    [tableColumns]
  ); // This results in an object like: { code: "", comment: "", ...etc }

  const { handleSubmit, register, unregister, control, watch } = useForm({
    defaultValues: {
      createEntities: [formFields],
    },
  });
  const { fields, append, remove } = useFieldArray({
    control,
    name: 'createEntities',
  });

// In my main render I have
         <button class="add-row" onClick={() => append(formFields, false)}> Add </button >
         {fields.map((item, index) => {
            const { id } = item;

            return (
              <div key={id}>
                <div class="row-number">
                    {index + 1}
                   <button onClick={() => remove(index)} ></button>
                </div>
                <Row
                    tableColumns={tableColumns}
                    columns={Object.keys(formFields)}
                    rowNumber={index}
                    register={register}
                />
              </StyledForm>
            );
          })}

// In the render from Row, which basically renders all inputs horizontally.

      {columns.map((columnKey, index) => {
        return (
          <div class="wrapper" key={index}>
            <FormField
              name={`createEntities[${rowNumber}].${columnKey}`}
              register={register}
             />
          </div>
        );
      })}

// And finally the FormField component renders the input

    <input
        type="text"
        name={name}
        defaultValue={value || ''}
        ref={register()}
        placeholder={placeholder || name}
      />

```

All this works for the basic stuff, rendering, updating the values for each row... the problem comes when I try to have let's say more than 10 rows, the rendering gets visibly and painfully slow. I tried adding React.memo to the FormField and to the Row components, but when I do that, if I try to remove rows from the middle (like removing 3 and then 5 from 1,2,3,4,5,6) the form breaks with an error on `const { id } = item` saying that can't get the id of undefined. 

Besides, I can see by printing the form data, that if I remove row 3 for example, the resulting form data gets spliced from that index instead of just removing the 3 from the array.

Any ideas on how to improve performance here? Or maybe I'm doing something terribly wrong?

Let me know if there's a need for more info.

Thanks!

**Comment by bluebill1049** [comment]
hey are you using `watch()` at the root of your form? can you try to provide a simple example with codesandbox?

**Reply by Geddard** [reply]
> hey are you using `watch()` at the root of your form? can you try to provide a simple example with codesandbox?

Hey there, thanks for the reply, I was using watch, which made me notice I had double the renders thanks to that. So I'll see if I can use the handleSubmit to get the data (although it was not working well on me before).

But I discovered that the culprit was actually some font-icons i was rendering next to each input... removing them boosts performance by like 1000% xD

**Comment by Geddard** [comment]
@bluebill1049 still, i think we have to consider the issue of the React.memo situation.

As I mentioned before, if i wrap the FormField in React.memo, the remove behavior breaks badly. And i think it is because the form hook is internally doing some magic that react.memo prevents, but it's not at all clear what exactly is going on.
